File: bindings/dotnet/Xui/SplitAnimation.cs
using System.Runtime.InteropServices;

namespace Xui;

public sealed partial class SplitView
{
    public uint TransitionDuration
    {
        get { Window.Guard(); Window.Check(Native.SplitDurationGet(Handle, out uint value)); return value; }
        set { Window.Guard(); Window.Check(Native.SplitDurationSet(Handle, value)); }
    }
    public float Progress
    {
        get { Window.Guard(); Window.Check(Native.SplitProgress(Handle, out float value)); return value; }
    }
    public bool Animating
    {
        get { Window.Guard(); Window.Check(Native.SplitAnimating(Handle, out uint value)); return value != 0; }
    }
    public SplitView SetTransitionDuration(uint milliseconds) { TransitionDuration = milliseconds; return this; }
}

internal static partial class Native
{
    [LibraryImport("xui", EntryPoint = "xui_split_view_set_transition_duration")]
    internal static partial int SplitDurationSet(ulong target, uint milliseconds);
    [LibraryImport("xui", EntryPoint = "xui_split_view_get_transition_duration")]
    internal static partial int SplitDurationGet(ulong target, out uint milliseconds);
    [LibraryImport("xui", EntryPoint = "xui_split_view_get_progress")]
    internal static partial int SplitProgress(ulong target, out float progress);
    [LibraryImport("xui", EntryPoint = "xui_split_view_get_animating")]
    internal static partial int SplitAnimating(ulong target, out uint animating);
}

File: bindings/rust/xui/src/split_animation.rs
use super::*;

impl SplitView {
    pub fn set_transition_duration(&self, milliseconds: u32) -> Result<()> {
        check(unsafe { sys::xui_split_view_set_transition_duration(self.handle, milliseconds) })
    }
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn split_animation_contract() -> Result<()> {
        let window = Window::new("Animated panes", 1000., 600.)?;
        let first = window.text_input("First")?;
        let second = window.text_input("Second")?;
        let split = window.split_view("Panes", &first, &second)?;
        assert_eq!(split.transition_duration()?, 0);
        assert_eq!(split.progress()?, 1.);
        assert!(!split.animating()?);
        split.set_second_visible(false)?;
        assert_eq!(split.progress()?, 0.);
        split.set_transition_duration(180)?;
        split.set_second_visible(true)?;
        assert!(split.animating()?);
        assert!(split.set_transition_duration(10001).is_err());
        assert_eq!(split.transition_duration()?, 180);
        assert!(split.animating()?);
        split.set_transition_duration(0)?;
        assert_eq!(split.progress()?, 1.);
        assert!(!split.animating()?);
        assert_ne!(
            unsafe { sys::xui_split_view_get_progress(split.handle, std::ptr::null_mut()) },
            0
        );
        let mut progress = 0.;
        assert_ne!(
            unsafe { sys::xui_split_view_get_progress(first.handle, &mut progress) },
            0
        );
        Ok(())
    }
}
impl SplitView {
    pub fn transition_duration(&self) -> Result<u32> {
        let mut value = 0;
        check(unsafe { sys::xui_split_view_get_transition_duration(self.handle, &mut value) })?;
        Ok(value)
    }
    pub fn progress(&self) -> Result<f32> {
        let mut value = 0.;
        check(unsafe { sys::xui_split_view_get_progress(self.handle, &mut value) })?;
        Ok(value)
    }
    pub fn animating(&self) -> Result<bool> {
        let mut value = 0;
        check(unsafe { sys::xui_split_view_get_animating(self.handle, &mut value) })?;
        Ok(value != 0)
    }
}
